impl/src/main/java/impl/QuestionResourceImpl.java

```diff
@@ -170,15 +198,38 @@ private static MarkdownTextObject markdownText(String string, String... args) {
         return questionDao.getQuestion(questionId)
             .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_QUESTION_FROM_DATABASE, throwable)))
             .switchIfEmpty(exception(() -> new WebException(NOT_FOUND, QUESTION_NOT_FOUND)))
-            .flatMap(storedQuestion -> {
+            .concatMap(storedQuestion -> {
                 if (auth.getUserId() != storedQuestion.getUserId()) {
                     return error(new WebException(FORBIDDEN, NOT_OWNER_OF_QUESTION));
                 }
                 return questionDao.updateQuestion(auth.getUserId(), questionId, question)
                     .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_UPDATE_QUESTION_TO_DATABASE, throwable)))
-                    .flatMap(ignore -> questionDao.getQuestion(questionId)
+                    .concatMap(ignore -> questionDao.getQuestion(questionId)
                         .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_QUESTION_FROM_DATABASE, throwable))))
-                    .switchIfEmpty(exception(() -> new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_QUESTION_FROM_DATABASE)));
+                    .switchIfEmpty(exception(() -> new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_QUESTION_FROM_DATABASE)))
+                    .concatMap(updatedQuestion -> {
+                        List<Observable<Integer>> daoCalls = new ArrayList<>();
+                        if(question.getTags() != null) { // Null means we shouldn't touch existing tags
+                            List<String> labels = question
+                                .getTags()
+                                .stream()
+                                .map(tag -> tag.getLabel().toLowerCase())
+                                .collect(Collectors.toList());
+                            for (Tag tag : question.getTags()) {
+                                daoCalls.add(tagDao.mergeTag(tag.getLabel()));
+                            }
+                            daoCalls.add(tagDao.removeTagAssociationFromQuestion(updatedQuestion.getId()));
```

Comment:
include updateQuestion in transaction , will make the use of getQuestion redundant